Audit the installed headers for C++ compatibility

MisterDA opened the pull request:
C++ name mangling applies to symbols: variable and function names. The rule of thumb is to enclose public symbols declarations in blocks, to avoid C++ linkers look for C++ symbol names instead of C.

```c
#ifdef __cplusplus
extern "C" {
#endif

/* symbols go here */

#ifdef __cplusplus
}
#endif
```

Symbols protected by `CAML_INTERNALS` blocks need not to be covered. I've sometimes reordered a few functions to merge two sections protected by `CAML_INTERNALS`.

Headers that contain definitions incompatible with C++, such as `_Atomic`, also need protection.

There shouldn't be anything but comments before header guards to allow for [multiple include optimization](https://gcc.gnu.org/onlinedocs/cppinternals/Guard-Macros.html).

Fix #13541. Close #11557.

MisterDA committed:
Audit the installed headers for C++ compatibility

C++ name mangling applies to symbols: variable and function names. The
rule of thumb is to enclose public symbols declarations in blocks:

    #ifdef __cplusplus
    extern "C" {
    #endif

    /* symbols go here */

    #ifdef __cplusplus
    }
    #endif

Symbols protected by CAML_INTERNALS blocks need not to be covered.

Headers that contain definitions incompatible with C++, such as
_Atomic, also need protection.

MisterDA committed:
Minor header inclusion fixes

NickBarnes commented on runtime/caml/osdeps.h:
```diff
@@ -139,12 +130,14 @@ CAMLextern int caml_win32_isatty(int fd);
 
 CAMLextern value caml_win32_get_temp_path(void);
 
+#include <time.h>
```
I'd be inclined to move this to the top of the `_WIN32` section, for consistency.

MisterDA replied:
Done, thanks.

NickBarnes reviewed (commented):
Looks good to me; consistency is a good thing.

MisterDA committed:
Test whether C++ stubs can be compiled and linked

- Test whether OCaml C headers are also valid in C++;
- Test whether C++ files can be linked with the runtime. Symbols that
  are not covered by CAML_INTERNALS need to have C linkage (under
  extern "C" { ... }). There are too many to check exhaustively, so
  use a dummy stub for now.

Co-authored-by: Kate <kit-ty-kate@outlook.com>

MisterDA commented:
I've added a test that checks [whether C++ stubs can be compiled and linked](https://github.com/ocaml/ocaml/pull/13591/commits/9ba790d284ff177a46f42ed0f2c9ed2e9b6305d8). I think it replaces and should close #11557. Thanks @kit-ty-kate for the tip!

This should be considered for 5.3 and 5.2.

ghost commented:
Mangling of variable names only matters when namespaces are used, which is obviously not the case in OCaml. Therefore several of the added guards are unnecessary, e.g. `runtime/caml/address_class.h` doesn't need any.

MisterDA commented:
> Therefore several of the added guards are unnecessary, e.g. `runtime/caml/address_class.h` doesn't need any.

Is that a strong blocker? Can we keep them for future use, or should I remove them?

ghost commented:
No, it's not a blocker.

Octachron commented:
@NickBarnes , did you have the time to finish your review? If possible I would like to include this fix in the next beta for 5.3 (that should happen next week hopefully).

MisterDA commented:
I'm tempted to add the C++ guards to all headers, unconditionally, even encompassing `CAML_INTERNALS` sections…

NickBarnes reviewed (approved):
LGTM. Having a test is great. As noted in your comment, it would be nice if the test's `#include`s were auto-generated, but that's a problem for another time.

NickBarnes commented:
@Octachron I apologise for the delay.

NickBarnes commented:
Merged and cherry-picked to 5.3.

Octachron commented:
I agree that cherry-picking 38b9c9e6a19b1c6bc57b6 to 5.3 is a good idea too to avoid diverging ocamltest tests.

NickBarnes commented:
Done under @dra27's close supervision ....
